### PR TITLE
Signature verification

### PR DESCRIPTION
**Pull request by mythmon**
I wrote some tests to verify that recipe signatures actually sign the data the recipe is supposed to be. This revealed a problem where recipes auto-signed on save were signing the wrong data. I fixed that bug.

r?

### CODE REVIEW

**Comment by mythmon** [comment]
@chartjes can you review the changes to the contract tests here?

**Comment by rehandalal** [comment]
i couldn't see any issues and circle is happy. r+ from me.

but i'll leave this open so @chartjes can review the contract tests.

File: package.json [hunk 1]
```diff
@@ -26,7 +26,7 @@
     "json-editor": "0.7.28",
     "json-loader": "0.5.4",
     "moment": "2.14.1",
-    "node-sass": "3.9.1",
+    "node-sass": "3.9.3",
```
Comment (Osmose) [inline package.json]: Unrelated?

File: normandy/recipes/utils.py [hunk 2]
```diff
@@ -115,3 +116,33 @@ def sign_data(self, content_list):
...
+
+
+def verify_signature(data, signature, pubkey):
+    if isinstance(data, str):
+        data = data.encode()
+
+    # add data template
+    data = b'Content-Signature:\x00' + data
+
+    signature = base64.b64decode(un_urlsafe(signature))
```
Comment (Osmose) [inline normandy/recipes/utils.py]: Could you just use `base64.urlsafe_b64decode` here?

File: normandy/recipes/utils.py [hunk 3]
```diff
@@ -115,3 +116,33 @@ def sign_data(self, content_list):
...
+        data,
+        hashfunc=hashlib.sha384,
+        sigdecode=ecdsa.util.sigdecode_string)
+
+    print(dir(verifying_pubkey))
+
+    return verified
+
+
+def un_urlsafe(input):
```
Comment (Osmose) [inline normandy/recipes/utils.py]: Could you document this function to explain what it's doing?

File: normandy/recipes/utils.py [hunk 4]
```diff
@@ -115,3 +116,33 @@ def sign_data(self, content_list):
...
+    pubkey = un_urlsafe(pubkey)
+
+    verifying_pubkey = ecdsa.VerifyingKey.from_pem(pubkey)
+    verified = verifying_pubkey.verify(
+        signature,
+        data,
+        hashfunc=hashlib.sha384,
+        sigdecode=ecdsa.util.sigdecode_string)
+
+    print(dir(verifying_pubkey))
```
Comment (Osmose) [inline normandy/recipes/utils.py]: A debug statement?

File: normandy/recipes/utils.py [hunk 5]
```diff
@@ -115,3 +116,33 @@ def sign_data(self, content_list):
...
+
+    return verified
+
+
+def un_urlsafe(input):
+    input = input.replace("_", "/")
+    input = input.replace("-", "+")
+    if len(input) % 4 > 0:
+        input += "=" * (4 - len(input) % 4)
```
Comment (Osmose) [inline normandy/recipes/utils.py]: Wait, we're decoding stuff but you still need to pad it? That seems odd to me.

File: normandy/recipes/utils.py (hunk 2 above)
Comment (rehandalal) [inline normandy/recipes/utils.py]: i thought this was doing the opposite of `urlsafe_b64decode` but maybe i'm wrong.

File: normandy/recipes/tests/test_models.py [hunk 6]
```diff
@@ -172,12 +173,30 @@ def test_update_signature(self, mocker):
         ]
 
         recipe = RecipeFactory(signed=False)
-        assert recipe.signature is None
```
Comment (Osmose) [inline normandy/recipes/tests/test_models.py]: So we no longer expect signatures to be empty at first?

File: normandy/recipes/utils.py (hunk 5 above)
Comment (mythmon) [inline normandy/recipes/utils.py]: Autograph is weird. It uses this not-quite-base64 thing that uses `_-` instead of `/+`, and doesn't include the padding, which we need for `base64.b64decode`. Maybe `base64.urlsafe_b64decode` will deal with all this for us, I'll give it shot. Otherwise I'll add a docstring to this function.

File: package.json (hunk 1 above)
Comment (mythmon) [inline package.json]: This was on master, and was needed to get the tests to pass. I can probably get rid of it with some rebasing.

File: normandy/recipes/utils.py (hunk 4 above)
Comment (mythmon) [inline normandy/recipes/utils.py]: Oops

File: normandy/recipes/tests/test_models.py (hunk 6 above)
Comment (mythmon) [inline normandy/recipes/tests/test_models.py]: Not when we have a working (or mocked) Autograph, since we automatically sign in `save()` now.

File: normandy/recipes/utils.py (hunk 5 above)
Comment (Osmose) [inline normandy/recipes/utils.py]: They _don't_ use padding? whyyyyyyyyyyyyyyyyyyyyyyy

**Comment by Osmose** [comment]
I'm r+wc for the non-contract-tests bits, I'm good with either removing the `un_urlsafe` function or documenting it. A test or two for it would be nice but not required if it stays.

**Comment by mythmon** [comment]
Turns out `un_urlsafe` isn't needed. We can use `base64.urlsafe_b64decode`, and ecdsa will automatically detect urlsafe b64, it seems. I'll be updating the Autograph code (which this is based on).

**Comment by chartjes** [comment]
Nicely done with the contract test updates -- it hadn't occurred to me to explicitly check that requests were okay, so raise_for_status() is a nice addition.
